darshan-util/pydarshan/darshan/report.py

```diff
@@ -241,8 +241,38 @@ def to_df(self, attach="default"):
         records = copy.deepcopy(self._records)
 
         if mod in ['LUSTRE']:
+            # retrieve the counter column names
+            c_cols = self.report.counters[mod]['counters']
+            # preallocate arrays to fill with the various record data
+            n_records = len(records)
+            ids = np.empty(n_records, dtype=np.uint64)
+            ranks = np.empty(n_records, dtype=np.int64)
+            counters = np.empty((n_records, len(c_cols)), dtype=np.int64)
+            # the OST ID's are jagged so they cannot be preallocated
+            ost_ids = []
             for i, rec in enumerate(records):
-                rec = rec
+                ids[i] = rec['id']
+                ranks[i] = rec['rank']
+                counters[i, :] = rec['counters']
+                ost_ids.append(rec['ost_ids'])
+
+            # create the counter dataframe and add a column for the OST ID's
+            counter_df = pd.DataFrame(counters, columns=c_cols)
+            counter_df["ost_ids"] = ost_ids
+
+            if attach:
+                if "id" in attach:
+                    # if the ids are attached, build a dataframe for them
+                    # then prepend it to the original expected dataframe
+                    id_df = pd.DataFrame(ids, columns=["id"])
+                    counter_df = pd.concat([id_df, counter_df], axis=1)
+                if "rank" in attach:
+                    # if the ranks are attached, prepend them as well
+                    rank_df = pd.DataFrame(ranks, columns=["rank"])
+                    counter_df = pd.concat([rank_df, counter_df], axis=1)
```

Comment:
Rather than having two different blocks of code that effectively do the same thing with `attach`, it is probably easier to abstract to a helper function. If you want to just reuse the old code initially and then change that to the new version for performance improvement on each code path later that's probably fine.

Since this PR isn't focused on performance improvement, and I haven't seen convincing performance arguments yet anyway, this also seems a fair bit more verbose than i.e., (the comments probably aren't needed for this level of simplicity either) the following.

```diff
             if attach:
                 if "id" in attach:
-                    # if the ids are attached, build a dataframe for them
-                    # then prepend it to the original expected dataframe
-                    id_df = pd.DataFrame(ids, columns=["id"])
-                    counter_df = pd.concat([id_df, counter_df], axis=1)
+                    counter_df.insert(0, "id", ids)
                 if "rank" in attach:
-                    # if the ranks are attached, prepend them as well
-                    rank_df = pd.DataFrame(ranks, columns=["rank"])
-                    counter_df = pd.concat([rank_df, counter_df], axis=1)
+                    counter_df.insert(0, "rank", ranks)
```

Reply:
Come to think of it, I'm not sure I see a justification for most of the code in use here; there's probably something even slicker/more performant, though I'm mostly thinking of avoiding iteration/embracing pandas.

```diff
diff --git a/darshan-util/pydarshan/darshan/report.py b/darshan-util/pydarshan/darshan/report.py
index 18d724a..3d6a698 100644
--- a/darshan-util/pydarshan/darshan/report.py
+++ b/darshan-util/pydarshan/darshan/report.py
@@ -243,33 +243,15 @@ class DarshanRecordCollection(collections.abc.MutableSequence):
         if mod in ['LUSTRE']:
             # retrieve the counter column names
             c_cols = self.report.counters[mod]['counters']
-            # preallocate arrays to fill with the various record data
-            n_records = len(records)
-            ids = np.empty(n_records, dtype=np.uint64)
-            ranks = np.empty(n_records, dtype=np.int64)
-            counters = np.empty((n_records, len(c_cols)), dtype=np.int64)
-            # the OST ID's are jagged so they cannot be preallocated
-            ost_ids = []
-            for i, rec in enumerate(records):
-                ids[i] = rec['id']
-                ranks[i] = rec['rank']
-                counters[i, :] = rec['counters']
-                ost_ids.append(rec['ost_ids'])
-
-            # create the counter dataframe and add a column for the OST ID's
-            counter_df = pd.DataFrame(counters, columns=c_cols)
-            counter_df["ost_ids"] = ost_ids
+            df_recs = pd.DataFrame(records)
+            counter_df = pd.DataFrame(df_recs.counters.tolist(), columns=c_cols)
+            counter_df["ost_ids"] = df_recs.ost_ids
 
             if attach:
                 if "id" in attach:
-                    # if the ids are attached, build a dataframe for them
-                    # then prepend it to the original expected dataframe
-                    id_df = pd.DataFrame(ids, columns=["id"])
-                    counter_df = pd.concat([id_df, counter_df], axis=1)
+                    counter_df.insert(0, "id", df_recs["id"])
                 if "rank" in attach:
-                    # if the ranks are attached, prepend them as well
-                    rank_df = pd.DataFrame(ranks, columns=["rank"])
-                    counter_df = pd.concat([rank_df, counter_df], axis=1)
+                    counter_df.insert(0, "rank", df_recs["rank"])
```

Reply:
Yeah, this is a pretty dramatic simplification. It looks like both cases are dominated by the `records = copy.deepcopy(self._records)` that occurs just above this, so they can't be that different in performance. I was already planning on switching to using `df.insert()` so I don't see a reason not to use this much simpler version.